Tasks/Maven/maven2.js

```diff
@@ -48,40 +50,161 @@ var publishJUnitResults = tl.getInput('publishJUnitResults');
+
+        if (!matchingTestResultsFiles) {
+            tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');
+            return 0;
+        }
+
+        var tp = new tl.TestPublisher("JUnit");
+        tp.publish(matchingTestResultsFiles, false, "", "");
+    }
+}
+
+function getSonarQubeRunner() {
+    var sqAnalysisEnabled = tl.getInput('sqAnalysisEnabled', true);
+
+    if (sqAnalysisEnabled != 'true') {
+        console.log("SonarQube analysis is not enabled.");
+        return;
+    }
+
+    console.log("SonarQube analysis is enabled");
+    var sqArguments;
+    var sqEndpoint = getEndpointDetails("sqConnectedServiceName");
+    var sqDbDetailsRequired = tl.getInput('sqDbDetailsRequired', true);
+
+    if (sqDbDetailsRequired == 'true') {
+        var sqDbUrl = tl.getInput('sqDbUrl', false);
+        var sqDbUsername = tl.getInput('sqDbUsername', false);
+        var sqDbPassword = tl.getInput('sqDbPassword', false);
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password, sqDbUrl, sqDbUsername, sqDbPassword);
+    } else {
+        sqArguments = createSonarQubeArgs(sqEndpoint.Url, sqEndpoint.Username, sqEndpoint.Password);
+    }
+
+    sqArguments = mavenOptions + " " + sqArguments;
+    tl.debug("Running Maven with goal sonar:sonar and options " + sqArguments);
+
+    var mvnsq = new tl.ToolRunner(mvntool);
+    mvnsq.arg('-f');
+    mvnsq.arg(mavenPOMFile);
+    mvnsq.arg(sqArguments);
+    mvnsq.arg("sonar:sonar");
+
+    return mvnsq;
+}
+
+function getEndpointDetails(inputFieldName) {
+    var errorMessage = "Could not decode the generic endpoint. Please ensure you are running the latest agent (min version 0.3.0)"
+    if (!tl.getEndpointUrl) {
+        throw new Error(errorMessage);
+    }
+
+    var genericEndpoint = tl.getInput(inputFieldName);
+    if (!genericEndpoint) {
+        throw new Error(errorMessage);
     }
-    else {
-      tl.debug('No pattern found in testResultsFiles parameter');
-      var matchingTestResultsFiles = [testResultsFiles];
+
+    hostUrl = tl.getEndpointUrl(genericEndpoint, false);
+    var auth = tl.getEndpointAuthorization(genericEndpoint, false);
+
+    if (auth.scheme != "UsernamePassword") {
+        throw new Error("The authorization scheme " + auth.scheme + " is not supported for a SonarQube endpoint. Please use a username and a password.");
     }
 
-    if(!matchingTestResultsFiles) {
-      tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');  
-      return 0;
+    hostUsername = auth.parameters.Username;
+    hostPassword = auth.parameters.Password;
+
+    return {
+        "Url": hostUrl,
+        "Username": hostUsername,
+        "Password": hostPassword
+    };
+}
+
+function createSonarQubeArgs(sqHostUrl, sqHostUsername, sqHostPassword, sqDbUrl, sqDbUsername, sqDbPassword) {
```

Comment:
Same as above, can this be replaced by adding each arg to the toolrunner directly?